Refactor library.

SFE-Brudnerd opened the pull request:
Full refactor of the library.
Works with SparkFun Toolkit.
Reduced complexity.
Added comments.

SFE-Brudnerd committed:
Delete

SFE-Brudnerd committed:
add files in.

SFE-Brudnerd committed:
WIP update.

Delete toolkit files.
Migrate registers file to header.
Refactor of library in progress.

SFE-Brudnerd committed:
Compiles but is untested.
Still need to finish commenting.

SFE-Brudnerd committed:
Added initialization defaults.

SFE-Brudnerd committed:
Get refactor working.

Update constructor defaults.
Add getter for device Address.
Add getters for outputs.
Edit comments for brevity.
Make fsr variables const instead of static constexpr.
Consolidate begin functions.
Add nullptr checks.
Incorporate new readRegisterRegion functions.
Rework prepareMeasurement.
Add comments.

SFE-Brudnerd committed:
Rework examples for updated library.

SFE-Brudnerd committed:
Rev library to 2.0.0.
Add dependency.

SFE-Brudnerd committed:
Add support for repeat starts in begin function

SFE-Brudnerd committed:
Get v2.0.0 ready for arduino library.

gigapod commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
+    /// @brief  Sets up Arduino I2C driver using the specified I2C address then calls the super class begin.
+    /// @param address Address of the I2C device.
+    /// @return True if successful, false otherwise.
+    bool begin(const uint8_t &address = kDefaultAS7331Addr)
```
A lot of our Arduino libraries have a begin that takes a Wire reference - should we do this here (maybe have a couple begins . This is minor.

sfe-SparkFro replied:
I'd argue this is not minor, some platforms just won't work with the default `Wire` (the Uno R4 comes to mind, it needs `Wire1` for the Qwiic connector). Can just add a second parameter that defaults to `Wire` like so:

```
bool begin(const uint8_t &address = kDefaultAS7331Addr, TwoWire& wirePort)
{
    if (_theI2CBus.init(wirePort, address) != kSTkErrOk)
        return false;

    // etc...
}
```

This is pretty typical of `begin()` in our libraries

SFE-Brudnerd replied:
That's easy enough to do. Just for clarification, did you mean to not include a default in the begin function?

```C++
bool begin(const uint8_t &address = kDefaultAS7331Addr, TwoWire &wirePort = Wire)
{
  ...
}
```

Or was this to be a second begin function?

```C++
bool begin(const uint8_t &address = kDefaultAS7331Addr)
{
  ...
}

bool begin(const uint8_t &address = kDefaultAS7331Addr, TwoWire &wirePort)
{
  ...
}
```

nseidle replied:
I think I echo @sfe-SparkFro, the user exposed begin() has, for the last 4+ years followed the following format:

* bool begin() - Nothing needed
* bool begin(alternate device address)
* bool begin(alternate device address, alternate Wire port)

Bool is ideally returned so the user can quickly check if the device is setup and online. If a more complex result can be returned, a different function should be made available. Rarely does the user need an alternate device address. *Very* rarely does the user need to modify the Wire port, but it's super nice to have when you need it.

SFE-Brudnerd replied:
Could that not be accomplished with defaults though?

```C++
bool begin(const uint8_t &address = kDefaultAS7331Addr, TwoWire &wirePort = Wire)
{
  ...
}
```

This could solve all 3 of those use cases. The only thing it wouldn't solve would be the user wanting to only set the Wire port but not the address, however I feel like that would be an extremely rare use case and the user can just enter the default address at that point.

sfe-SparkFro replied:
Yes sorry, meant to have the default port in the params: `TwoWire &wirePort = Wire`

Agreed, if the user needs to specify the port, it's not exactly hard for them to specify the address too.

nseidle replied:
Yup, your solution works, I just want to be sure we don't change things.

gigapod commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
+};
 
-            sfe_as7331_reg_cfg_optreg_t optreg;
+class SfeAS7331ArdI2C : public SfeAS7331Driver
```
We probably need to standardize the naming pattern of the Arduino driver implementations - I know they are all over the place right now.  This is fine though - but something for our future selfs to hash a bit (I'm not sure if Arduino has any style guidelines here - ...).

gigapod reviewed (approved):
Overall looks solid - glad it's working!

Thanks for MAJOR refactor work.

sfe-SparkFro commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
+    /// @brief Called to reset all local values back to initial conditions.
+    void setDefaultSettings(void);
+
+    sfeTkIBus *_theBus;  // Pointer to bus device.
```
I'd argue that because the AS7331 can only communicate over I2C, this should be a `sfeTkII2C`

SFE-Brudnerd replied:
I'm not too worried about this.

sfe-SparkFro commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
-        static constexpr float fsrC = 169984.0; // Full Scale Resolution for the UVC sensor.
+    /// @brief Checks to see if the AS7331 is connected.
+    /// @return True if successful, false otherwise.
+    bool isConnected(void)
```
Related to my other comment about changing `sfeTkIBus *_theBus` to a `sfeTkII2C`, this should be moved into `SfeAS7331Driver`. This method would be identical on all platforms, so it should be in the base class.

SFE-Brudnerd replied:
In this specific instance, sure. However, from a standardization perspective this should exist in the interface-specific implementation. Since this is supposed to be one of the first examples for proper implementation of the Toolkit library, we should do it correctly for a generic implementation.

sfe-SparkFro commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
-            _config.indexMode = indexMode;
+        if (!reset())
+            return false;
 
-            return SFE_BUS_OK;
-        }
+        isConnected();
```
Checking if the device is connected and resetting it should not be done here, they should be done as part of `SfeAS7331Driver::begin()`. The platform-specific `begin()` should only do things that are exclusive to the platform.

SFE-Brudnerd replied:
Going back to comments from the previous implementation, you can't make assumptions about the operating mode of the device on reset, so you need to make sure it's in the right mode. The easiest way to do this is to reset the device to make sure it's in a known configuration. This is standard practice.

sfe-SparkFro replied:
I understand that the operating mode needs to be correct to do stuff, but that's not my point. The point is that resetting the device and checking if it's connected (whatever order that happens in) is independent of what platform you're on, so they should be done within the base class to reduce duplicate code and keep the scope of things where they should be.

To your other point about `isConnected()` being implementation-specific - just the act of calling `getDeviceID()` effectively also pings the device, because if it's not on the bus, then that creates an error that gets handled differently. But if you want to be "proper" about it, then you could add a virtual `isConnected()` in the base class, and implement that in the derived classes. Then you can still move all the code for resetting and checking if it's connected into the base class, which will reduce duplicate code and keep the scope of things where they should be.

SFE-Brudnerd replied:
I removed the reset, so this isn't really valid anymore. Now that getDeviceID will move it to cfg mode and back it wasn't necessary. 

However, getting the deviceID is different from ping and is interface specific.

I don't agree that by adding a second isConnected and doing half in there and half in the implementation specific would in any way reduce duplicate code. If anything it would increase it.

sfe-SparkFro commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
+    /// @brief Sets the UV sensor's gain.
+    /// @param gain The gain to set the sensor to.
+    /// @return 0 if successful, negative if error, positive for warning.
+    sfeTkError_t setGain(const as7331_gain_t &gain);
```
I'm curious, what's the motivation for making almost every parameter pass by reference? And why are a couple parameters not pass by reference? (namely the in constructor, `begin`, `setCommunicationBus`, `runDefaultSetup`, and `prepareMeasurement`)

SFE-Brudnerd replied:
Reduces copying of existing data. Constructor could be passed by reference. `begin` does besides the bus. `setCommunicationBus` and `begin` use a pointer because they're setting an internal pointer. References are checked at compile time, and the constructor doesn't define a default Bus to use. `runDefaultSetup` could use a reference, so I'll do that. It should also be const. Same with `prepareMeasurement`.

sfe-SparkFro commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
+};
 
-            sfe_as7331_reg_cfg_optreg_t optreg;
+class SfeAS7331ArdI2C : public SfeAS7331Driver
```
I'd be in favor of having the platform-specific code in a separate file from the generic class. So the Arduino stuff stays in this file (because of how Arduino does includes), then the generic code moves into another file. Makes it easier for anyone who needs to port to a different platform, they don't have to remember to delete the Arduino stuff.

SFE-Brudnerd replied:
Good point, I can do that. I was more thinking if you were to add say a Zephyr implementation that you would just add that to the bottom. Only compiles what it uses.

sfe-SparkFro replied:
Yes, it technically only compiles what it uses, but this file currently wouldn't be able to compile at all for another platform (eg. Zephyr) due o the inclusion of `#include <Arduino.h>` (and vice versa if other platforms are added). Each platform implementation needs to go in its own file

sfe-SparkFro commented on src/SparkFun_AS7331.cpp:
```diff
@@ -0,0 +1,1051 @@
+uint8_t SfeAS7331Driver::getDeviceID(void)
+{
+    // Config registers are only available in Configuration mode.
+    if (!_theBus || _opMode != DEVICE_MODE_CFG)
+        return 0;
```
Rather than doing nothing, I'd suggest switching modes, read the device ID, then switch back (assuming that's not too hard, right?). This is a more special method that won't be called often, but when it is called, I really think it should do as requested.

SFE-Brudnerd replied:
I could see that being helpful. It will be a bit more involved since on reset the default internal state could be different than the device's actual state. I agree that this is a special case and I can work around that.

sfe-SparkFro commented on src/SparkFun_AS7331.cpp:
```diff
@@ -0,0 +1,1051 @@
+
+void SfeAS7331Driver::setDeviceAddress(const uint8_t &deviceAddress)
+{
+    _devAddress = deviceAddress;
```
Should verify that `deviceAddress` is actually valid for this device (ie. 0x74 through 0x77)

sfe-SparkFro commented on src/SparkFun_AS7331.cpp:
```diff
@@ -0,0 +1,1051 @@
+{
+    // Config registers are only available in Configuration mode.
+    if (!_theBus || _opMode != DEVICE_MODE_CFG)
+        return false;
```
This is another special method that I think should do as requested, even if it's in the wrong mode. In fact, you don't even need to change back at the end if it starts in the wrong mode.

sfe-SparkFro commented on src/SparkFun_AS7331.cpp:
```diff
@@ -0,0 +1,1051 @@
+        return false;
+
+    // Calculate new conversion factors to make sure they match the current settings.
+    calculateConversionFactors();
```
I think the majority of this code (lines 68-123) should only be executed in the `else` case, right? If `reset()` is called, all the registers will already be at their default values, so writing them to their default values again is pointless.

sfe-SparkFro commented on src/SparkFun_AS7331.cpp:
```diff
@@ -0,0 +1,1051 @@
+    return true;
+}
+
+bool SfeAS7331Driver::reset(void)
+{
+    // Standard read-modify-write sequence.
+    sfe_as7331_reg_cfg_osr_t osr;
+
+    if (kSTkErrOk != getOSR(osr))
+        return false;
+
+    // Set software reset bit.
+    osr.sw_res = 1;
+
+    if (kSTkErrOk != setOSR(osr))
+        return false;
```
Should check `_opMode` first and change mode if needed

SFE-Brudnerd replied:
OSR is available in all modes. This is commented down in `getOSR` but I can add a note here I suppose.

sfe-SparkFro replied:
Ah, gotcha, all good then!

sfe-SparkFro commented on src/SparkFun_AS7331.h:
```diff
@@ -14,1312 +14,716 @@
+    /// @param divider Divider value to set.
+    /// @param setEnableDiv Option to turn on the divider if desired.
+    /// @return 0 if successful, negative if error, positive for warning.
+    sfeTkError_t setDigitalDividerRange(const as7331_divider_val_t &divider, const bool &enableDiv = false);
```
`enableDiv` should really default to true. If I call `mySensor.setDigitalDividerRange(DIV_n)`, I'd expect that to actually do something.

sfe-SparkFro commented on src/SparkFun_AS7331.cpp:
```diff
@@ -0,0 +1,1051 @@
+    return kSTkErrOk;
+}
+
+sfeTkError_t SfeAS7331Driver::getOSR(sfe_as7331_reg_cfg_osr_t &osrReg)
+{
+    // OSR is available in both operation modes.
+    if (!_theBus)
+        return kSTkErrFail;
+
+    return _theBus->readRegisterByte(kSfeAS7331RegCfgOsr, osrReg.byte);
+}
+
+sfeTkError_t SfeAS7331Driver::setOSR(const sfe_as7331_reg_cfg_osr_t &osrReg)
+{
+    // OSR is available in both operation modes.
+    if (!_theBus)
+        return kSTkErrFail;
+
+    return _theBus->writeRegisterByte(kSfeAS7331RegCfgOsr, osrReg.byte);
+}
```
Shouldn't these check `_opMode` like the other methods below?

SFE-Brudnerd replied:
Comments on lines 804/813 give the answer, but OSR is available in both modes.

sfe-SparkFro replied:
Oh man, I should really learn to read lol

sfe-SparkFro commented on src/SparkFun_AS7331.cpp:
```diff
@@ -0,0 +1,1051 @@
+        regAddress = kSfeAS7331RegMeasMres1;
+        fsr = &_fsrA;
+        retUV = &_uva;
+        if (_mmode != MEAS_MODE_SYND)
+        {
+            conv = &_conversionA;
+        }
```
Nit: no need to check `_mmode` here (and B and C), it gets checked later. Could save a whole 9 lines of code!

SFE-Brudnerd replied:
Good point, I was thinking not to assign if it if not necessary as that saves clock cycles, but in reality it should only be one clock cycle. The SYND mode uses a completely different variable.

sfe-SparkFro replied:
The comparison also takes time, so not comparing will actually save some time on average

sfe-SparkFro commented on examples/Example01_Basic_OneShot/Example01_Basic_OneShot.ino:
```diff
@@ -43,36 +45,35 @@ void setup() {
   
   // Wait for a bit longer than the conversion time.
-  delay(2+(1 << myUVSensor.getConfigConversionTime()));
+  delay(2+(1 << myUVSensor.getConversionTime()));
```
I think it'd be good to add a helper method `getConversionTimeMillis()` (or whatever name) to help the user. It's not immediately obvious from this code why the bitshifting is needed, and some folks are going to assume `getConversionTime()` gives them milliseconds.

In fact, it might help to rename it to `getConversionTimeRaw()` to help further distinguish them.

Related - it might be good to do something similar with any methods that return enums corresponding to "real" values, including `getGain()`, `getCClk()`, and `getDigitalDividerRange()`. That is, add `Raw` to the end of those method names, and add helper methods for each: `getGainScaler()`, `getCClkMHz()`, and `getDigitalDividerRangeScaler()` (idk about `Scaler`, names are hard lol)

SFE-Brudnerd replied:
We haven't really done this in any other library when you pull it out, but I definitely see the niceness of that. However, those functions aren't really used in the library at all, I just use the member variable directly. Think it's worth providing both raw and human-readable versions public to the user?

sfe-SparkFro replied:
> We haven't really done this in any other library

Actually, I've seen a mix within our libraries. Some return raw register values, some return "real" values ([example here](
https://github.com/sparkfun/SparkFun_MMC5983MA_Magnetometer_Arduino_Library/blob/8b22bc4e68e6f2331d419b0571dd93adc7a36f2f/src/SparkFun_MMC5983MA_Arduino_Library.cpp#L651)). It's not always clear when methods will return raw or "real" values, but in my mind, I sort of default to "real" values unless the method has `Raw` or something in the name.

> Think it's worth providing both...?

I'm guessing everyone's gonna have a different opinion on this, but I'd vote yes.

SFE-Brudnerd replied:
I'm not opposed to providing both.

sfe-SparkFro commented on examples/Example01_Basic_OneShot/Example01_Basic_OneShot.ino:
```diff
@@ -22,16 +22,18 @@
-
 void setup() {
   Serial.begin(115200);
   while(!Serial){delay(100);};
```
Have you tested this on various platforms? I recall `while(!Serial)` doing weird things sometimes, but maybe that's outdated.

Also good to just test the whole library on other platforms to confirm everything is happy.

SFE-Brudnerd replied:
Tested it on ESP32, nRF, and Atmel platforms. I've never had an issue with this tbh.

sfe-SparkFro replied:
Sounds good!

sfe-SparkFro commented on examples/Example01_Basic_OneShot/Example01_Basic_OneShot.ino:
```diff
@@ -43,36 +45,35 @@ void setup() {
   Serial.println("Sensor began.");
```
Couple comments on `myDevice.begin()` in general:

I've always found it much more helpful to have a look that continuously attempts to begin the device (1x per second is fine, don't go crazy), and prints an error message each time. I know, you shouldn't hotplug devices, but people are going to do it anyways, and this just saves them a step of having to hit the reset button (or re-uploading code if they don't have a reset button (yes, really, some dev boards don't)). And then there's much less chance of the user missing the single error message if they didn't have their serial monitor open or something.

If you insist on freezing, then change "Spinning" to "Freezing" or "Halting" or something that's more obvious (think about non-native English speakers).

SFE-Brudnerd replied:
I do insist on freezing, it's either going to be connected or it won't. I don't want to encourage bad behavior. I'll rename it to halting.

sfe-SparkFro commented on examples/Example02_CONT_Mode/Example02_CONT_Mode.ino:
```diff
@@ -51,14 +53,14 @@ void setup() {
 
   // Set break time to 900us (112 * 8us) to account for the time it takes to poll data.
-  if(SFE_BUS_OK != myUVSensor.setBreakTime(112)) {
+  if(kSTkErrOk != myUVSensor.setBreakTime(112)) {
```
I know the break time is explained at the top of the file, but I think it would help to reiterate it here as well for anyone who misses it at the top (like me 😉). Also explain why 900us is chosen

SFE-Brudnerd replied:
...silly pebkacs...

SFE-Brudnerd replied:
But also...

> Also explain why 900us is chosen

`to account for the time it takes to poll data.`

sfe-SparkFro replied:
> to account for the time it takes to poll data.

What I mean is to explain that part, because 900 is just a magic number without an explanation. Something like:

"A total of 10 bytes are read from the AS7331, split between 2 separate read transactions. Thus a total of 12 bytes are transmitted (including the address byte twice), or 108 bits (including +1 ACK bit per byte). At 100kHz, this takes at least 1080us to complete, so we set the break time to <insert reasonable number of microseconds that should be longer than 1080us to account for other things taking time>"

And as described above, you actually need more than 900us

SFE-Brudnerd replied:
I think this is completely unnecessary. I don't need to account for every byte. I measured it and set it to that number experimentally.

Also where did you get "across 2 separate read transactions"? The `readAllUV` function only does one read. Additionally, the amount of time would vary across platforms, this was a best guess estimate. There's no need to explain that verbosely.

sfe-SparkFro replied:
If you measured it experimentally, then that should be stated in the comment. Eg. "This 900us was experimentally determined to be a suitable amount of time with a 100kHz clock frequency"

> Also where did you get "across 2 separate read transactions"?

My mistake, was thinking it also called `readOutConv()` which reads 4 more bytes, but that's only in `MEAS_MODE_SYND`.

Also, I still got the number of bytes wrong before (oops!), because each read operation actually has to write the register address first. So to read N bytes (6 in this case), a total of N+3 bytes (9 in this case) go across the bus. Working through the timings again, that should theoretically be 810ua at 100kHz, which is in line with your measurement. I still think it's worth offering this explanation in the comment so a user understands _why_ that much time is chosen beyond just experimental findings; it means if they change something (eg. clock frequency or going into `MEAS_MODE_SYND`), then they would know how to change the timing without needing to measure it experimentally.

sfe-SparkFro commented on examples/Example03_SYNS_Mode/Example03_SYNS_Mode.ino:
```diff
@@ -58,7 +60,7 @@ void setup() {
 
   // Set measurement mode and change device operating mode to measure.
-  if(myUVSensor.startMeasurement(MEAS_MODE_SYNS) == false) {
+  if(myUVSensor.prepareMeasurement(MEAS_MODE_SYNS) == false) {
```
I think it'd help to add a comment to explain that `MEAS_MODE_SYNS` is "SYNchronous Start mode", like in the `as7331_meas_mode_t` definition. It'd be easy for a user to miss the difference between it and `MEAS_MODE_SYND`

SFE-Brudnerd replied:
It's explained in the top comment description.

sfe-SparkFro commented on examples/Example04_SYND_Mode/Example04_SYND_Mode.ino:
```diff
@@ -51,7 +53,7 @@ void setup() {
 
   // Set measurement mode and change device operating mode to measure.
-  if(myUVSensor.startMeasurement(MEAS_MODE_SYND) == false) {
+  if(myUVSensor.prepareMeasurement(MEAS_MODE_SYND) == false) {
```
I think it'd help to add a comment to explain that `MEAS_MODE_SYND` is "SYNchronous start and enD mode", like in the `as7331_meas_mode_t` definition. It'd be easy for a user to miss the difference between it and `MEAS_MODE_SYNS`

SFE-Brudnerd replied:
It's explained in the top comment description.

sfe-SparkFro reviewed (changes requested):
Overall looks really good!

SFE-Brudnerd committed:
Code Review update.
Update Examples to say halting rather than spinning.
Add comments about why break time is set.
Move driver to it's own underlying file.
Make getDeviceID change op mode if needed.
Add checking to setDeviceAddress.
Make runDefaultSetup change op mode if needed.
Move bulk of runDefaultSetup into the else statement so we don't do unnecessary things.
prepareMeasurement now uses references and consts.
Add comment to reset about OSR register availability.
Add human readable versions of getters for gain, cclk, convTime.
Rename original versions of those to Raw.
Remove unnecessary if's in readRawUV.
setDigitalDividerRange now defaults enabling the divider to true.
Modify Arduino begin to take in a wire port.
Remove unnecessary reset call in begin due to logic changes above.

sfe-SparkFro commented on src/SparkFun_AS7331.h:
```diff
@@ -696,19 +37,16 @@ class SfeAS7331ArdI2C : public SfeAS7331Driver
-        if (!reset())
-            return false;
-
         isConnected();
```
Should `isConnected()` still be here? It's not doing anything currently.

SFE-Brudnerd replied:
Yes it should. It pings and then reads the ID to verify that it's connected before continuing the begin process. I don't understand how you think it doesn't do anything.

sfe-SparkFro replied:
Sorry, "not doing anything" was the wrong choice of words. I meant your code isn't checking the return value of `isConnected()`. If it returns false, then it should not continue.

SFE-Brudnerd replied:
I can get behind that.

sfe-SparkFro commented on src/sfeAS7331.cpp:
```diff
@@ -0,0 +1,1107 @@
+        return 0;
+
+    // If we changed it at first, change it back.
+    if(needsToBeChangedBack)
+    {
+        if(kSTkErrOk != setOperationMode(DEVICE_MODE_MEAS))
+            return 0;
+    }
```
Can save 1 read operation by instead doing:

```
osr.dos = DEVICE_MODE_MEAS;

if(kSTkErrOk != setOSR(osr))
    return 0;
```

SFE-Brudnerd replied:
Ah I fixed it above but forgot to fix it there. Thanks.

sfe-SparkFro commented on src/sfeAS7331.cpp:
```diff
@@ -0,0 +1,1107 @@
+            _devAddress = deviceAddress;
+            break;
+        default: // Otherwise set it to the default. No disallowed addresses.
+            _devAddress = kDefaultAS7331Addr;
```
I think the default case should just do nothing, and leave the address at whatever it was.

SFE-Brudnerd committed:
Update changes.
Make default case for setDeviceAddress do nothing.
Manually setOSR in getDeviceID.

SFE-Brudnerd committed:
update keywords.txt for new functions.

SFE-Brudnerd committed:
Merge branch 'main' into work

gigapod commented:
A lot of comments/commits - if you need me to look at anything just let me know...

SFE-Brudnerd commented:
I think we're all set now.
